### ISSUE TITLE
[css-highlight-api] Figure out how highlights are exposed to the accessibility tree

### ISSUE THREAD

**Issue by frivoal** [w3c/csswg-drafts#6498]
@atanassov recently reminded us that figure out how highlights are exposed to the accessibility tree. This had been know for a while, but we didn't have an issue I could find, so here's one.

**Comment by frivoal** [comment]
To sighted users, the highlights are only visible through the way they are styled. There is no other way to know what they're for. Maybe the same should be true of the highlights? Getting the highlight itself described through some text sounds like it might be more convenient, but I'm not sure where we'd pull that description from, how to deal with multiple languages, or the potential discrepancies between the descriptive text and what the highlight is actually used for.

**Comment by astearns** [comment]
And whatever gets decided here should likely match what is decided for #6417

**Comment by frivoal** [comment]
How are things like these exposed to the accessibility tree today?
* `::spelling-error`
* `::grammar-error`
* `::target-text`

We don't necessarily have to do the same, but it'd be good to start with an understanding of how these are handled.

**Comment by ffiori** [comment]
#4601 seems to have some related remarks.

**Comment by dlibby-** [comment]
UserAgent detected spelling and grammar errors (i.e. ::spelling-error and ::grammar-error) in Chromium do have a mapping to various AT APIs. These are serialized as part of the accessibility tree and as far as I can tell, this is not part of any of the Aria-related standards. For example on UIA on Windows spelling errors are exposed as a text range with an AnnotationType_SpellingError. Not sure how this would fit into the spec, but perhaps is a path forward (following the UIA example, there is an AnnotationType_Highlight).

For completeness, spelling and grammar errors that are computed by the webpage can be annotated with [aria-invalid](https://www.w3.org/TR/wai-aria/#aria-invalid). The corresponding concept for highlights would probably be the mark role (see https://w3c.github.io/aria/). However, these require additional DOM nodes, which is counter to the purpose of the HighlightAPI.

**Comment by dandclark** [comment]
I agree with @dlibby that for the question of how to expose the highlight ranges to ATs we should be able to follow the lead of how ::spelling-error and ::grammar-error are exposed in Chromium, at least in part.

For expressing the semantics of the highlight ranges, the most flexible option seems to be to add a `title` string attribute to `Highlight`, which would be exposed to ATs. However, as pointed out in https://github.com/w3c/csswg-drafts/issues/4601#issuecomment-806426095, these strings wouldn't be localized. This approach also depends on the author to provide a string with the correct level of descriptiveness.

An alternative I'd like to consider is adding an enum to `Highlight` that captures the common use cases we expect for this API. These would be exposed as appropriate in different accessibility platforms, e.g. in UIA the most closely matching [`AnnotationType`](https://docs.microsoft.com/en-us/windows/win32/winauto/uiauto-annotation-type-identifiers) would be used.

```c++
enum HighlightType {
 "highlight",
 "spellingerror", // Exposed to ATs in the same way as ::spelling-error
 "grammarerror", // Exposed to ATs in the same way as ::grammar-error
 "textmatch", // e.g. for custom find-on-page results
// ...and any others for the common use cases we can think of.
};

[Exposed=Window]
interface Highlight {
  constructor(AbstractRange... initialRanges);
  setlike<AbstractRange>;
  attribute long priority;
  attribute HighlightType type; // Exposed to ATs. Default is "highlight".
};
```

The benefit of this is that we're not relying on author-specified strings, so it's easier for developers and it solves the issue of localization. An enumerated set of types could also be easier to expose to ATs.

The disadvantage is that use cases we don't anticipate won't have any good choices so authors will have to fall back to the generic "highlight". I don't think we'd be locked into the initial set of HighlightTypes though; we could probably add more if additional common uses cases arise.

**Comment by michael-n-cooper** [comment]
Tagging for a11y-tracker, since the dup that was closed was so tagged.

**Comment by dandclark** [comment]
Another thing I'll add here is that an advantage of an enum over a string is that when OSs have some first-class representation for any of the use cases in the enum, user agents could plug into them directly. For example, custom spell check implementations using Highlight API could be represented to accessibility tools in the exact same way as ranges marked by the native spell-checker. In UIA this might [AnnotationType_SpellingError](https://docs.microsoft.com/en-us/windows/win32/winauto/uiauto-annotation-type-identifiers), or  [NSAccessibilityMarkedMisspelledTextAttribute](https://developer.apple.com/documentation/appkit/nsaccessibilitymarkedmisspelledtextattribute) on Mac.  This would be a more consistent experience for AT users and would save the author the trouble of localizing a "spelling error" string.

It's also feasible that in the future ATs could become more powerful and offer first-class representations of more of these use cases, and it would be great if authors can provide metadata that provides a better experience in those cases.

If needed, in an L2 of the spec a string attribute could be added to `Highlight` for use cases that fall outside of those we've expressed in the enum. But for use cases that become common and have support in ATs across OSs, it seems more powerful to enable these as additions to the enum. Allowing arbitrary strings could miss opportunities to line up with things that OSs recognize.

**Comment by dandclark** [comment]
There's been some delay on agenda space opening up to discuss this during CSSWG calls, so I wanted try again to see if we could make some progress here in the issue.

Considering the enum approach considered [here](https://github.com/w3c/csswg-drafts/issues/6498#issuecomment-908755397) and looking more at the current capabilities of UIA's [Annotation Type Identifiers](https://docs.microsoft.com/en-us/windows/win32/winauto/uiauto-annotation-type-identifiers) and Mac's [Attributed String Keys](https://developer.apple.com/documentation/foundation/nsattributedstring/key#:~:text=of%20the%20text.-,Accessibility%20Keys,-static%20let%20accessibilityAlignment), I'd like to propose starting with an enum with just these:

```js
enum HighlightType {
 // Exposed with AnnotationType_Highlight for UIA, accessibilityCustomText for Mac
 "highlight",
 // Exposed with AnnotationType_SpellingError for UIA, NSAccessibilityMisspelledTextAttribute for Mac
 "spellingerror",
 // Exposed with AnnotationType_GrammarError for UIA, NSAccessibilityMisspelledTextAttribute for Mac
 "grammarerror"
 // More can be added as use cases emerge and accessibility platforms add more text semantic types
};

[Exposed=Window]
interface Highlight {
  constructor(AbstractRange... initialRanges);
  setlike<AbstractRange>;
  attribute long priority;
  attribute HighlightType type; // Exposed to ATs. Default is "highlight".
};
```

This approach lets custom spellcheck achieve the same experience with accessibility tools as native spellcheck for UIA and Mac today, and leaves room to add more entries to the enum as use cases an support emerge. A custom string could be added in an L2 if the need for that level of flexibility emerges.

@frivoal @sanketj @dlibby- @lilles, any thoughts on this approach?

**Comment by dandclark** [comment]
>  I wonder if the generic highlight item in the enum could have a secondary field for color?

@SuzanneTaylor If the highlight range is being used to change the text color or background color of a range of text, the author will have had to include a rule for that in the `::highlight()` pseudo-element selector that the highlight is using. So if we want to include color in the information given to ATs about highlights, I think we could just reuse that style information. Allowing color to be specified separately on the `Highlight` interface could result in confusion if the color is changed in the CSS and the value on the interface becomes out-of-sync with the value in the `::highlight()` pseudo.

**Comment by maryjom** [comment]
For the use case of highlights in application-specific highlighting. Using this CSS API can an author define multiple different highlights and provide the associated text meaning of a particular color of highlight that can be relayed to AT using the CSS attributes? As an example, an app that scans through large amounts of text to highlight certain aspects particular to the search (e.g. to point out sentiments or emotions in text, such as positive, negative, neutral, happy, angry, etc.)

**Comment by SuzanneTaylor** [comment]
> 
> 
> > I wonder if the generic highlight item in the enum could have a secondary field for color?
> 
> @SuzanneTaylor If the highlight range is being used to change the text color or background color of a range of text, the author will have had to include a rule for that in the `::highlight()` pseudo-element selector that the highlight is using. So if we want to include color in the information given to ATs about highlights, I think we could just reuse that style information. Allowing color to be specified separately on the `Highlight` interface could result in confusion if the color is changed in the CSS and the value on the interface becomes out-of-sync with the value in the `::highlight()` pseudo.

That makes sense, but how will colors specified as numbers become human-friendly words? With contrast requirements already restricting color choices for the highlights, background and text colors, I don't think being limited to the named colors will work well.

**Comment by SuzanneTaylor** [comment]
Would it be possible also to indicate in any documentation about browser accessibility support for this feature, that browsers should consider a setting (or maybe there should be a media query for this) that users could select to have the name of the color appear visually? 

You would not want ALL CSS colors to appear visually of course, just those associated with highlights. One possible visual layout could be smaller text above the highlighted text, overlapping onto the highlight a bit.

**Comment by dandclark** [comment]
I think it's a good idea to at least add a note in the spec that browsers should ensure that the colors of the highlights are made available to accessibility.

I'm not yet sure whether there's a need for a setting to expose the highlight colors in a special way. In many (most?) cases, the specific color of the highlights won't matter semantically, e.g. with custom find-on-page, spelling/grammar error, and highlights used for generic emphasis of text. For the rarer cases where color matters, users would know from context on the page that the colors are meaningful and would be able to adjust their accessibility tool's settings accordingly to get that info.

That said I'd like to continue exploring an option like this in the L2 of the spec when we've got more experience with how the API ends up being used in the wild.

**Comment by MoritzGlantz** [comment]
Hey! I was wondering if this is actually working in any screen reader as of right now?

**Comment by ffiori** [comment]
@MoritzGlantz I've recently been working on expanding AX support for custom highlights on different platforms on Edge and chromium browsers. This is the current state of things for different highlight types (screen readers announce highlights for the green tick cases):

| Platform              | Custom | Spelling | Grammar |
|-----------------------|--------|----------|---------|
| Windows (UIA and IA2) | ✅     | ✅       | ✅      |
| Mac                   | ✅     | ✅       | ❌      |
| Linux                 | ❌     | ✅       | ✅      |

The Windows row includes Narrator, NVDA and JAWS. And I've been told there's current work ongoing for the two missing cases for Mac and Linux.
NVDA on Firefox seems to announce the three different types as well.
Unsure about support from Safari or other browsers.
Hope this helps.

**Comment by MoritzGlantz** [comment]
Hi @ffiori and thanks for posting this!

I can confirm it sometimes gets announced in NVDA + Chrome. JAWS does not announce anything on my machine (Windows 11 ARM, JAWS 2026, default settings). VoiceOver on macOS (Safari) also does not announce anything on my machine. Is this dependent on specific settings?

This is the demo I built for this: https://codepen.io/MoritzGlantz/pen/yyaVPoQ
